--- mediawiki-1.9.3/includes/LogPage.php	2007-02-20 20:20:32.000000000 -0600
+++ wiki19/includes/LogPage.php	2008-01-21 13:24:24.000000000 -0600
@@ -134,6 +134,9 @@
 		global $wgLang, $wgContLang, $wgLogActions;
 
 		$key = "$type/$action";
+        if( $key == 'patrol/patrol' )
+	        return PatrolLog::makeActionText( $title, $params, $skin );
+
 		if( isset( $wgLogActions[$key] ) ) {
 			if( is_null( $title ) ) {
 				$rv=wfMsg( $wgLogActions[$key] );
