--- mediawiki-1.9.3/includes/SpecialNewpages.php	2007-02-20 20:20:31.000000000 -0600
+++ wiki19/includes/SpecialNewpages.php	2008-02-13 07:03:56.000000000 -0600
@@ -158,7 +158,8 @@
  * constructor
  */
 function wfSpecialNewpages($par, $specialPage) {
-	global $wgRequest, $wgContLang;
+	global $wgRequest, $wgContLang, $wgOut;
+	$wgOut->setRobotPolicy("noindex,follow");
 
 	list( $limit, $offset ) = wfCheckLimits();
 	$namespace = NS_MAIN;
