--- mediawiki-1.9.3/includes/templates/Userlogin.php	2007-02-20 20:20:31.000000000 -0600
+++ wiki19/includes/templates/Userlogin.php	2008-01-28 13:20:40.000000000 -0600
@@ -31,13 +31,11 @@
 	<h2><?php $this->msg('login') ?></h2>
 	<p id="userloginlink"><?php $this->html('link') ?></p>
 	<div id="userloginprompt"><?php  $this->msgWiki('loginprompt') ?></div>
-	<?php if( @$this->haveData( 'languages' ) ) { ?><div id="languagelinks"><p><?php $this->html( 'languages' ); ?></p></div><?php } ?>
 	<table>
 		<tr>
 			<td align='right'><label for='wpName1'><?php $this->msg('yourname') ?>:</label></td>
 			<td align='left'>
 				<input type='text' class='loginText' name="wpName" id="wpName1"
-					tabindex="1"
 					value="<?php $this->text('name') ?>" size='20' />
 			</td>
 		</tr>
@@ -45,7 +43,6 @@
 			<td align='right'><label for='wpPassword1'><?php $this->msg('yourpassword') ?>:</label></td>
 			<td align='left'>
 				<input type='password' class='loginPassword' name="wpPassword" id="wpPassword1"
-					tabindex="2"
 					value="<?php $this->text('password') ?>" size='20' />
 			</td>
 		</tr>
@@ -58,8 +55,7 @@
 		<tr>
 			<td align='right'><?php $this->msg( 'yourdomainname' ) ?>:</td>
 			<td align='left'>
-				<select name="wpDomain" value="<?php $this->text( 'domain' ) ?>"
-					tabindex="3">
+				<select name="wpDomain" value="<?php $this->text( 'domain' ) ?>">
 					<?php echo $doms ?>
 				</select>
 			</td>
@@ -69,23 +65,20 @@
 			<td></td>
 			<td align='left'>
 				<input type='checkbox' name="wpRemember"
-					tabindex="4"
 					value="1" id="wpRemember"
-					<?php if( $this->data['remember'] ) { ?>checked="checked"<?php } ?>
+					checked="checked"
 					/> <label for="wpRemember"><?php $this->msg('remembermypassword') ?></label>
 			</td>
 		</tr>
 		<tr>
 			<td></td>
 			<td align='left' style="white-space:nowrap">
-				<input type='submit' name="wpLoginattempt" id="wpLoginattempt" tabindex="5" value="<?php $this->msg('login') ?>" />&nbsp;<?php if( $this->data['useemail'] && $this->data['canreset']) { ?><input type='submit' name="wpMailmypassword" id="wpMailmypassword"
-					tabindex="6"
+				<input type='submit' name="wpLoginattempt" id="wpLoginattempt" value="<?php $this->msg('login') ?>" />&nbsp;<?php if( $this->data['useemail'] ) { ?><input type='submit' name="wpMailmypassword" id="wpMailmypassword"
 									value="<?php $this->msg('mailmypassword') ?>" />
 				<?php } ?>
 			</td>
 		</tr>
 	</table>
-<?php if( @$this->haveData( 'uselang' ) ) { ?><input type="hidden" name="uselang" value="<?php $this->text( 'uselang' ); ?>" /><?php } ?>
 </form>
 </div>
 <div id="loginend"><?php $this->msgWiki( 'loginend' ); ?></div>
@@ -96,6 +89,7 @@
 
 class UsercreateTemplate extends QuickTemplate {
 	function execute() {
+			
 		if( $this->data['message'] ) {
 ?>
 	<div class="<?php $this->text('messagetype') ?>box">
@@ -106,28 +100,43 @@
 	</div>
 	<div class="visualClear"></div>
 <?php } ?>
+<style type='text/css'>
+.hiderow {
+	display:none;
+}
+
+</style>
+<script type='text/javascript'>
+	function show_hide_real_name() {
+		row = document.getElementById('real_name_row');
+		if (document.userlogin2.wpUseRealNameAsDisplay.checked) {
+			row.className='';
+		} else {
+			row.className = 'hiderow'
+			document.userlogin2.wpRealName.value = '';
+		}
+	}	
+</script>
+
 <div id="userlogin">
 
 <form name="userlogin2" id="userlogin2" method="post" action="<?php $this->text('action') ?>">
 	<h2><?php $this->msg('createaccount') ?></h2>
 	<p id="userloginlink"><?php $this->html('link') ?></p>
 	<?php $this->html('header'); /* pre-table point for form plugins... */ ?>
-	<?php if( @$this->haveData( 'languages' ) ) { ?><div id="languagelinks"><p><?php $this->html( 'languages' ); ?></p></div><?php } ?>
 	<table>
 		<tr>
 			<td align='right'><label for='wpName2'><?php $this->msg('yourname') ?>:</label></td>
 			<td align='left'>
 				<input type='text' class='loginText' name="wpName" id="wpName2"
-					tabindex="1"
-					value="<?php $this->text('name') ?>" size='20' />
+					value="<?php $this->text('name') ?>" size='20' tabindex='2'/>
 			</td>
 		</tr>
 		<tr>
 			<td align='right'><label for='wpPassword2'><?php $this->msg('yourpassword') ?>:</label></td>
 			<td align='left'>
 				<input type='password' class='loginPassword' name="wpPassword" id="wpPassword2"
-					tabindex="2"
-					value="<?php $this->text('password') ?>" size='20' />
+					value="<?php $this->text('password') ?>" size='20' tabindex='3'/>
 			</td>
 		</tr>
 	<?php if( $this->data['usedomain'] ) {
@@ -139,8 +148,7 @@
 		<tr>
 			<td align='right'><?php $this->msg( 'yourdomainname' ) ?>:</td>
 			<td align='left'>
-				<select name="wpDomain" value="<?php $this->text( 'domain' ) ?>"
-					tabindex="3">
+				<select name="wpDomain" value="<?php $this->text( 'domain' ) ?>">
 					<?php echo $doms ?>
 				</select>
 			</td>
@@ -150,28 +158,34 @@
 			<td align='right'><label for='wpRetype'><?php $this->msg('yourpasswordagain') ?>:</label></td>
 			<td align='left'>
 				<input type='password' class='loginPassword' name="wpRetype" id="wpRetype"
-					tabindex="4"
-					value="<?php $this->text('retype') ?>"
+					value="<?php $this->text('retype')?>" tabindex='4'
 					size='20' />
 			</td>
 		</tr>
 		<tr>
 			<?php if( $this->data['useemail'] ) { ?>
-				<td align='right'><label for='wpEmail'><?php $this->msg('youremail') ?></label></td>
+				<td align='right'><label for='wpEmail'><?php $this->msg('youremail') ?>:</label></td>
 				<td align='left'>
 					<input type='text' class='loginText' name="wpEmail" id="wpEmail"
-						tabindex="5"
-						value="<?php $this->text('email') ?>" size='20' />
+						value="<?php $this->text('email') ?>" size='20' tabindex='5'/>
 				</td>
 			<?php } ?>
 			<?php if( $this->data['userealname'] ) { ?>
 				</tr>
 				<tr>
-					<td align='right'><label for='wpRealName'><?php $this->msg('yourrealname') ?></label></td>
+				<tr>
+					<td colspan=2><input type='checkbox' id='wpUseRealNameAsDisplay' name='wpUseRealNameAsDisplay' onchange='show_hide_real_name();' tabindex='6'/><label for="wpUseRealNameAsDisplay">
+					<?php $this->msg('user_real_name_display'); ?>
+					</label>
+				</td>
+				</tr>
+				
+				<tr id='real_name_row' class='hiderow'>
+
+					<td align='right'><label for='wpRealName'><?php $this->msg('yourrealname') ?>:</label></td>
 					<td align='left'>
 						<input type='text' class='loginText' name="wpRealName" id="wpRealName"
-							tabindex="6"
-							value="<?php $this->text('realname') ?>" size='20' />
+							value="<?php $this->text('realname') ?>" size='20' tabindex='7'/>
 					</td>
 			<?php } ?>
 		</tr>
@@ -179,9 +193,7 @@
 			<td></td>
 			<td align='left'>
 				<input type='checkbox' name="wpRemember"
-					tabindex="7"
-					value="1" id="wpRemember"
-					<?php if( $this->data['remember'] ) { ?>checked="checked"<?php } ?>
+					value="1" id="wpRemember" checked="checked" tabindex='8'
 					/> <label for="wpRemember"><?php $this->msg('remembermypassword') ?></label>
 			</td>
 		</tr>
@@ -189,40 +201,27 @@
 			<td></td>
 			<td align='left'>
 				<input type='submit' name="wpCreateaccount" id="wpCreateaccount"
-					tabindex="8"
-					value="<?php $this->msg('createaccount') ?>" />
+					value="<?php $this->msg('createaccount') ?>" tabindex='9'/>
 				<?php if( $this->data['createemail'] ) { ?>
 				<input type='submit' name="wpCreateaccountMail" id="wpCreateaccountMail"
-					tabindex="9"
-					value="<?php $this->msg('createaccountmail') ?>" />
+					value="<?php $this->msg('createaccountmail') ?>" tabindex='10'/>
 				<?php } ?>
 			</td>
 		</tr>
 	</table>
 	<?php
 
-		if ($this->data['userealname'] || $this->data['useemail']) {
-			echo '<div id="login-sectiontip">';
-			if ( $this->data['useemail'] ) {
-				echo '<div>';
-				$this->msgHtml('prefs-help-email');
-				echo '</div>';
-			}
-			if ( $this->data['userealname'] ) {
-				echo '<div>';
-				$this->msgHtml('prefs-help-realname');
-				echo '</div>';
-			}
+		if( $this->data['useemail'] ) {
+			echo '<div id="login-emailforlost">';
+			$this->msgWiki( 'emailforlost' );
 			echo '</div>';
 		}
 
 	?>
-<?php if( @$this->haveData( 'uselang' ) ) { ?><input type="hidden" name="uselang" value="<?php $this->text( 'uselang' ); ?>" /><?php } ?>
 </form>
 </div>
 <div id="signupend"><?php $this->msgWiki( 'signupend' ); ?></div>
 <?php
-
 	}
 }
 
